File: common/src/main/java/juuxel/adorn/platform/FluidBridge.java
package juuxel.adorn.platform;

import juuxel.adorn.fluid.FluidAmountPredicate;
import juuxel.adorn.fluid.FluidUnit;
import juuxel.adorn.fluid.FluidVolume;
import juuxel.adorn.util.InlineServices;
import juuxel.adorn.util.Services;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.material.Fluid;
import org.jspecify.annotations.Nullable;

@InlineServices
public interface FluidBridge {
    FluidUnit getFluidUnit();

    @Nullable FluidVolume drain(Level world, BlockPos pos, @Nullable BlockState state, Direction side, Fluid fluid, FluidAmountPredicate amountPredicate);

    @InlineServices.Getter
    static FluidBridge get() {
        return Services.load(FluidBridge.class);
    }
}

File: common/src/main/kotlin/juuxel/adorn/platform/FluidBridge.kt
package juuxel.adorn.platform

import juuxel.adorn.fluid.FluidAmountPredicate
import juuxel.adorn.fluid.FluidUnit
import juuxel.adorn.fluid.FluidVolume
import juuxel.adorn.util.InlineServices
import juuxel.adorn.util.loadService
import net.minecraft.block.BlockState
import net.minecraft.fluid.Fluid
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.world.World

interface FluidBridge {
    val fluidUnit: FluidUnit

    fun drain(world: World, pos: BlockPos, state: BlockState?, side: Direction, fluid: Fluid, amountPredicate: FluidAmountPredicate): FluidVolume?

    @InlineServices
    companion object {
        private val instance: FluidBridge by lazy { loadService() }

        @JvmStatic
        fun get(): FluidBridge = instance
    }
}
